# Naive Bayes Klasifikasi

**Nama  : Rizki Ardian Samudra**

**NIM   : 210411100179**

**Kelas : Penambangan Data-C**

Naive Bayes pada dasarnya adalah sebuah algoritma klasifikasi yang berdasarkan pada teorema Bayes dengan mengasumsikan bahwa setiap fitur (atau atribut) pada data yang diamati adalah independen satu sama lain. Oleh karena itu, algoritma ini dianggap "naif" karena tidak mempertimbangkan interaksi antara fitur.

Dalam klasifikasi, Naive Bayes menghitung probabilitas setiap kelas yang mungkin berdasarkan pada nilai fitur yang diamati, dan kemudian memilih kelas dengan probabilitas tertinggi sebagai prediksi. Algoritma ini sering digunakan dalam analisis teks, seperti pengklasifikasian email spam atau klasifikasi dokumen berdasarkan topik, tetapi juga dapat diterapkan pada berbagai jenis masalah klasifikasi lainnya.

1. import library pandas, numpy, dan sklearn yang diperlukan

In [ ]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

2. import data kita

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/datasets/breast-cancer/master/data/breast-cancer.csv')

jumlah_baris = df.shape[0]
df.head(jumlah_baris)

age mefalsepause tumor-size inv-falsedes falsede-caps  deg-malig  \
0    40-49   premefalse      15-19          0-2         True          3   
1    50-59         ge40      15-19          0-2        False          1   
2    50-59         ge40      35-39          0-2        False          2   
3    40-49   premefalse      35-39          0-2         True          3   
4    40-49   premefalse      30-34          3-5         True          2   
..     ...          ...        ...          ...          ...        ...   
267  50-59         ge40      30-34          6-8         True          2   
268  50-59   premefalse      25-29          3-5         True          2   
269  30-39   premefalse      30-34          6-8         True          2   
270  50-59   premefalse      15-19          0-2        False          2   
271  50-59         ge40      40-44          0-2        False          3   

    breast breast-quad  irradiat                    class  
0    right     left_up     False        recurrence-events  
1    right     central     False  false-recurrence-events  
2     left    left_low     False        recurrence-events  
3    right    left_low      True  false-recurrence-events  
4     left    right_up     False        recurrence-events  
..     ...         ...       ...                      ...  
267   left    left_low     False  false-recurrence-events  
268   left    left_low      True  false-recurrence-events  
269  right    right_up     False  false-recurrence-events  
270  right    left_low     False  false-recurrence-events  
271   left    right_up     False  false-recurrence-events  

[272 rows x 10 columns]

3. Ubah nama fitur yang memiliki strip (-)

In [ ]:
df = df.rename(columns={'tumor-size': 'tumor_size'})
df = df.rename(columns={'inv-falsedes': 'inv_falsedes'})
df = df.rename(columns={'falsede-caps': 'falsede_caps'})
df = df.rename(columns={'deg-malig': 'deg_malig'})
df = df.rename(columns={'breast-quad': 'breast_quad'})
df = df.rename(columns={'class': 'class_name'})
df.head(jumlah_baris)

age mefalsepause tumor_size inv_falsedes falsede_caps  deg_malig  \
0    40-49   premefalse      15-19          0-2         True          3   
1    50-59         ge40      15-19          0-2        False          1   
2    50-59         ge40      35-39          0-2        False          2   
3    40-49   premefalse      35-39          0-2         True          3   
4    40-49   premefalse      30-34          3-5         True          2   
..     ...          ...        ...          ...          ...        ...   
267  50-59         ge40      30-34          6-8         True          2   
268  50-59   premefalse      25-29          3-5         True          2   
269  30-39   premefalse      30-34          6-8         True          2   
270  50-59   premefalse      15-19          0-2        False          2   
271  50-59         ge40      40-44          0-2        False          3   

    breast breast_quad  irradiat               class_name  
0    right     left_up     False        recurrence-events  
1    right     central     False  false-recurrence-events  
2     left    left_low     False        recurrence-events  
3    right    left_low      True  false-recurrence-events  
4     left    right_up     False        recurrence-events  
..     ...         ...       ...                      ...  
267   left    left_low     False  false-recurrence-events  
268   left    left_low      True  false-recurrence-events  
269  right    right_up     False  false-recurrence-events  
270  right    left_low     False  false-recurrence-events  
271   left    right_up     False  false-recurrence-events  

[272 rows x 10 columns]

4. Menangani data hilang atau missing value

> dibawah ini menghitung data kosong:

In [ ]:
df_copy = df.copy()
df_copy.isnull().sum()

age             0
mefalsepause    0
tumor_size      0
inv_falsedes    0
falsede_caps    8
deg_malig       0
breast          0
breast_quad     1
irradiat        0
class_name      0
dtype: int64

> ganti nilai yang mengandung tanda tanya (?) menjadi nilai kosong

In [ ]:
df_copy = df_copy.replace('?',np.nan)

In [ ]:
df_copy.isnull().sum()

age             0
mefalsepause    0
tumor_size      0
inv_falsedes    0
falsede_caps    8
deg_malig       0
breast          0
breast_quad     1
irradiat        0
class_name      0
dtype: int64

5. Cek type data dari setiap fitur

In [ ]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272 entries, 0 to 271
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   age           272 non-null    object
 1   mefalsepause  272 non-null    object
 2   tumor_size    272 non-null    object
 3   inv_falsedes  272 non-null    object
 4   falsede_caps  264 non-null    object
 5   deg_malig     272 non-null    int64 
 6   breast        272 non-null    object
 7   breast_quad   271 non-null    object
 8   irradiat      272 non-null    bool  
 9   class_name    272 non-null    object
dtypes: bool(1), int64(1), object(8)
memory usage: 19.5+ KB


In [ ]:
df_copy['deg_malig'] = df_copy['deg_malig'].astype('object')
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272 entries, 0 to 271
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   age           272 non-null    object
 1   mefalsepause  272 non-null    object
 2   tumor_size    272 non-null    object
 3   inv_falsedes  272 non-null    object
 4   falsede_caps  264 non-null    object
 5   deg_malig     272 non-null    object
 6   breast        272 non-null    object
 7   breast_quad   271 non-null    object
 8   irradiat      272 non-null    bool  
 9   class_name    272 non-null    object
dtypes: bool(1), object(9)
memory usage: 19.5+ KB


6. Ringkasan statistik dari semua atribut:

In [ ]:
df_copy.describe(include= 'all')

age mefalsepause tumor_size inv_falsedes falsede_caps  deg_malig  \
count     272          272        272          272          264        272   
unique      6            3         11            7            2          3   
top     50-59   premefalse      30-34          0-2        False          2   
freq       91          143         60          200          209        124   

       breast breast_quad irradiat               class_name  
count     272         271      272                      272  
unique      2           5        2                        2  
top      left    left_low    False  false-recurrence-events  
freq      143         103      205                      191

7. Langkah Imputasi Data atau mengisi data hilang dengan nilai yang didasarkan top dari ringkasan statistik diatas

In [ ]:
for value in ['falsede_caps', 'breast_quad']:
    df_copy[value].replace([None], [df_copy.describe(include='all')[value][2]], inplace=True)  #mengisi berdasarkan top describe

df_copy.head(jumlah_baris)

age mefalsepause tumor_size inv_falsedes  falsede_caps deg_malig  \
0    40-49   premefalse      15-19          0-2          True         3   
1    50-59         ge40      15-19          0-2         False         1   
2    50-59         ge40      35-39          0-2         False         2   
3    40-49   premefalse      35-39          0-2          True         3   
4    40-49   premefalse      30-34          3-5          True         2   
..     ...          ...        ...          ...           ...       ...   
267  50-59         ge40      30-34          6-8          True         2   
268  50-59   premefalse      25-29          3-5          True         2   
269  30-39   premefalse      30-34          6-8          True         2   
270  50-59   premefalse      15-19          0-2         False         2   
271  50-59         ge40      40-44          0-2         False         3   

    breast breast_quad  irradiat               class_name  
0    right     left_up     False        recurrence-events  
1    right     central     False  false-recurrence-events  
2     left    left_low     False        recurrence-events  
3    right    left_low      True  false-recurrence-events  
4     left    right_up     False        recurrence-events  
..     ...         ...       ...                      ...  
267   left    left_low     False  false-recurrence-events  
268   left    left_low      True  false-recurrence-events  
269  right    right_up     False  false-recurrence-events  
270  right    left_low     False  false-recurrence-events  
271   left    right_up     False  false-recurrence-events  

[272 rows x 10 columns]

8. Pastikan sudah tidak ada data yang kosong lagi

In [ ]:
df_copy.isnull().sum()

age             0
mefalsepause    0
tumor_size      0
inv_falsedes    0
falsede_caps    0
deg_malig       0
breast          0
breast_quad     0
irradiat        0
class_name      0
dtype: int64

9. One-Hot Encoder

> Encoder ini mengonversi data ke dalam format biner.

> Berikut adalah kode untuk menggunakan LabelEncoder pada beberapa kolom categorical pada dataset adult_df, dan kemudian melakukan encoding menggunakan One-Hot Encoder:

In [ ]:
le = preprocessing.LabelEncoder()
age = le.fit_transform(df_copy.age)
mefalsepause = le.fit_transform(df_copy.mefalsepause)
tumor_size = le.fit_transform(df_copy.tumor_size)
inv_falsedes = le.fit_transform(df_copy.inv_falsedes)
falsede_caps = le.fit_transform(df_copy.falsede_caps)
deg_malig = le.fit_transform(df_copy.deg_malig)
breast = le.fit_transform(df_copy.breast)
breast_quad = le.fit_transform(df_copy.breast_quad)
irradiat = le.fit_transform(df_copy.irradiat)
class_name = le.fit_transform(df_copy.class_name)

In [ ]:
#initialize the encoded categorical columns
df_copy['age'] = age
df_copy['mefalsepause'] = mefalsepause
df_copy['tumor_size'] = tumor_size
df_copy['inv_falsedes'] = inv_falsedes
df_copy['falsede_caps'] = falsede_caps
df_copy['deg_malig'] = deg_malig
df_copy['breast'] = breast
df_copy['breast_quad'] = breast_quad
df_copy['irradiat'] = irradiat
df_copy['class_name'] = class_name

df_copy.head(jumlah_baris)

age  mefalsepause  tumor_size  inv_falsedes  falsede_caps  deg_malig  \
0      2             2           2             0             1          2   
1      3             0           2             0             0          0   
2      3             0           6             0             0          1   
3      2             2           6             0             1          2   
4      2             2           5             4             1          1   
..   ...           ...         ...           ...           ...        ...   
267    3             0           5             5             1          1   
268    3             2           4             4             1          1   
269    1             2           5             5             1          1   
270    3             2           2             0             0          1   
271    3             0           7             0             0          2   

     breast  breast_quad  irradiat  class_name  
0         1            2         0           1  
1         1            0         0           0  
2         0            1         0           1  
3         1            1         1           0  
4         0            4         0           1  
..      ...          ...       ...         ...  
267       0            1         0           0  
268       0            1         1           0  
269       1            4         0           0  
270       1            1         0           0  
271       0            4         0           0  

[272 rows x 10 columns]

10. Melakukan Standardisasi Data atau Standardization of Data

> Semua nilai data dari kerangka data adalah numerik. Sekarang, kita perlu mengonversinya dalam satu skala. Kita bisa menstandarkan nilai. Kita dapat menggunakan rumus di bawah ini untuk standardisasi:

In [ ]:
num_features = ['age', 'mefalsepause', 'tumor_size', 'inv_falsedes', 'falsede_caps', 'deg_malig', 'breast', 'breast_quad', 'irradiat']

scaled_features = {}
for each in num_features:
    mean, std = df_copy[each].mean(), df_copy[each].std()
    scaled_features[each] = [mean, std]
    df_copy.loc[:, each] = (df_copy[each] - mean)/std

11. Pemotongan Data atau Data Slicing

> Pisahkan data menjadi data training and data test. Kita dapat dengan mudah melakukan langkah ini menggunakan metode train_test_split() dari sklearn:

> Variabel features_train berisi data fitur untuk training, features_test berisi data fitur untuk test, target_train berisi target untuk training, dan target_test berisi target untuk test.

In [ ]:
features = df_copy.values[:,:9]  #Variabel features diisi dengan nilai features dari dataset df_copy dengan slicing [:,:9] artinya mengambil seluruh baris dan kolom 0 hingga kolom 3 (kolom ke-9 tidak diambil karena merupakan variabel target).
target = df_copy.values[:,9]  #Variabel target diisi dengan nilai target dari dataset df_copy dengan slicing [:,9] artinya mengambil seluruh baris dan hanya kolom ke-9 (variabel target).
features_train, features_test, target_train, target_test = train_test_split(features,target, test_size = 0.33, random_state = 10)  #Kemudian menggunakan fungsi train_test_split, dataset yang sudah dipisahkan dijadikan argumen, dengan tambahan argumen test_size = 0.33 yang berarti dataset akan dibagi menjadi 33% data uji dan 67% data latih, dan random_state = 10 sebagai seed untuk menghasilkan hasil yang sama setiap kali dijalankan.

12. Implementasi Gaussian Naive Bayes

> Setelah selesai preprocessing data. saatnya menerapkan algoritma pembelajaran mesin atau machine learning. Kita akan menggunakan modul GaussianNB sklearn:

In [ ]:
clf = GaussianNB()
clf.fit(features_train, target_train)
target_pred = clf.predict(features_test)

13. Akurasi model Gaussian Naive Bayes

In [ ]:
accuracy = accuracy_score(target_test, target_pred, normalize = True)

print(f"Akurasi model KNN: {accuracy:.2%}")

Akurasi model KNN: 76.67%


## Kesimpulan

Berdasarkan nilai diatas tingkat akurasinya 76.67% . Ini menunjukkan data kita bagus. Kita dapat membuat kumpulan data uji acak dan menguji model untuk mengetahui seberapa baik kinerja model Gaussian Naive Bayes yang terlatih.